File: Data Structures & Algorithms/products-of-array-discluding-self/submission-6.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [1]*n

        prefix_val = 1
        for i in range(1, n):
            prefix_val *= nums[i-1]
            result[i] = prefix_val

        postfix_val = 1
        for i in range(n-2, -1, -1):
            postfix_val *= nums[i+1]
            result[i] *= postfix_val
            
        return result

File: Data Structures & Algorithms/top-k-elements-in-list/submission-6.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> um;
        
        for(auto num:nums){
            um[num]++;
        }
        vector<int> init{};
        vector<vector<int>> buckets(nums.size()+1,init);

        for(auto itr=um.begin();itr!=um.end();itr++){
            buckets[itr->second].push_back(itr->first);
        }

        vector<int> result{};
        for(int idx=buckets.size()-1;idx>0;idx--){
                for(auto val: buckets[idx]){
                    result.push_back(val);
                    if (result.size()==k) return result;
                }
            }
        return result;

    }
};
